src/imageops/transform.rs

```diff
@@ -0,0 +1,64 @@
+extern crate rayon;
+use self::rayon::prelude::*;
+
+use decoders::{OrientationFlips, RawImage};
+use imageops::OpBuffer;
+
+/// Mirror an OpBuffer horizontally
+pub fn flip_horizontal(buf: &OpBuffer) -> OpBuffer {
+  let mut out = OpBuffer::new(buf.width, buf.height, buf.colors);
+  out.data.par_chunks_mut(out.width * out.colors).enumerate().for_each(|(row, line)| {
+    let offset = buf.width * row * buf.colors;
+    for col in 0 .. buf.width {
+      for c in 0 .. buf.colors {
+        line[col * buf.colors + c] = buf.data[offset + (buf.width - 1 - col) * buf.colors + c];
+      }
+    }
+  });
+
+  out
+}
+
+/// Mirror an OpBuffer vertically
+pub fn flip_vertical(buf: &OpBuffer) -> OpBuffer {
+  let mut out = OpBuffer::new(buf.width, buf.height, buf.colors);
+  out.data.par_chunks_mut(out.width * out.colors).enumerate().for_each(|(row, line)| {
+    let offset = (buf.height - 1 - row) * buf.width * buf.colors;
+    for col in 0 .. buf.width * buf.colors {
+      line[col] = buf.data[offset + col];
+    }
+  });
+
+  out
+}
+
+/// Transpose an OpBuffer
+pub fn transpose(buf: &OpBuffer) -> OpBuffer {
+  let mut out = OpBuffer::new(buf.height, buf.width, buf.colors);
+
+  out.data.par_chunks_mut(out.width * out.colors).enumerate().for_each(|(row, line)| {
+    for col in 0 .. buf.height {
+      let target = col * buf.colors;
+      let source = (col * buf.width + row) * buf.colors;
+      for c in 0 .. buf.colors {
+        line[target + c] = buf.data[source + c];
+      }
+    }
+  });
+
+  out
+}
+
+/// Rotate an OpBuffer based on the given RawImage's orientation
+pub fn rotate(img: &RawImage, buf: &OpBuffer) -> OpBuffer {
+  match img.orientation.into() {
+    OrientationFlips { transpose: false, flip_x: false, flip_y: false } => buf.clone(),
+    OrientationFlips { transpose: false, flip_x: false, flip_y: true } => flip_vertical(buf),
+    OrientationFlips { transpose: false, flip_x: true, flip_y: false } => flip_horizontal(buf),
+    OrientationFlips { transpose: false, flip_x: true, flip_y: true } => flip_horizontal(&flip_vertical(buf)),
+    OrientationFlips { transpose: true, flip_x: false, flip_y: false } => transpose(buf),
+    OrientationFlips { transpose: true, flip_x: false, flip_y: true } => flip_vertical(&transpose(buf)),
+    OrientationFlips { transpose: true, flip_x: true, flip_y: false } => flip_horizontal(&transpose(buf)),
+    OrientationFlips { transpose: true, flip_x: true, flip_y: true } => flip_vertical(&transpose(&flip_horizontal(buf))),
+  }
+}
```

Comment:
I have done them once for my first unfinished pet project (https://github.com/runfalk/rustyspans/tree/master/src), but I'm not really sold on the currently recommended method of having them in the same file they are testing. I'm more for stuffing them in a separate file and/or folder to avoid unnecessary clutter. If you think you have a good idea of how to structure the tests, feel free to do it post merge.

I made a rough draft idea for how to test `OpBuffer` in a human readable way: https://gist.github.com/runfalk/7dda18b548d6b58fc59892e89648aa3b